test/com/facebook/buck/maven/PublisherIntegrationTest.java

```diff
@@ -87,4 +108,67 @@ public void testPublishFiles() throws Exception {
     assertThat(putRequestsInvoked, hasItem(String.format(urlTemplate, "pom")));
     assertThat(putRequestsInvoked, hasItem(String.format(urlTemplate, "pom.sha1")));
   }
+
+  @Test
+  public void shouldInvokeSignerIfRequested()
+      throws IOException, NoSuchBuildTargetException, InterruptedException, DeploymentException {
+    ProjectFilesystem filesystem = new ProjectFilesystem(temp.getRoot());
+
+    MavenPublishable publishable = (MavenPublishable)
+        JavaLibraryBuilder.createBuilder(BuildTargetFactory.newInstance("//:lib#maven"), filesystem)
+            .setMavenCoords("com.example:lib:1.0")
+            .addSrc(new FakeSourcePath("Foo.java"))
+            .build(new BuildRuleResolver(
+                TargetGraph.EMPTY,
+                new DefaultTargetNodeToBuildRuleTransformer()));
+    // Ensure that the output has been created, as if we'd run the build
+    filesystem.createParentDirs(publishable.getPathToOutput());
+    filesystem.touch(publishable.getPathToOutput());
+
+    // We also need to have the signature file because we can't know that gpg is available. By the
+    // time the publisher is run, we expect a set of output files to have been created. Each
+    // output file will have a ".asc" file generated by invoking the process executor. Set
+    // everything up.
+    Set<String> signatures = new HashSet<>();
+    ImmutableMap.Builder<ProcessExecutorParams, FakeProcess> commands = ImmutableMap.builder();
+    ImmutableSet<String> outputs = ImmutableSet.of(
+        publishable.getPathToOutput().toString(),
+        "buck-out/gen/lib#maven.pom");
+    for (String out : outputs) {
+      Path path = filesystem.getPathForRelativePath(out);
+      Path signed = filesystem.getPathForRelativePath(out + ".asc");
+
+      commands.put(
+          ProcessExecutorParams.builder().addCommand(
+          "gpg", "-ab", "--batch", "--passphrase", "passphrase", path.toString()).build(),
+          new FakeProcess(0, "", ""));
+      signatures.add(signed.toString());
+
+      filesystem.touch(path);
+    }
+    FakeProcessExecutor executor = new FakeProcessExecutor(commands.build());
+
+    Publisher publisher = new Publisher(
+        filesystem,
+        Optional.of(temp.newFolder().toUri().toURL()),
+        Optional.empty(),
+        Optional.empty(),
+        Optional.of("passphrase"),
+        false) {
+      @Override
+      protected ProcessExecutor createProcessExecutor(PrintStream stdout, PrintStream stderr) {
+        try {
+          for (String signature : signatures) {
+            Path path = filesystem.getRootPath().getFileSystem().getPath(signature);
+            filesystem.touch(path);
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        return executor;
+      }
+    };
+
+    publisher.publish(ImmutableSet.of(publishable));
+  }
```

Comment:
If the `.asc` is missing and signing has been enabled, the check at `Publisher.java:322` will cause an exception to be thrown. The test will fail if the executor isn't called, and I've added an assert that every expected command is executed.